src/modules/MagneticFieldReader/MagneticFieldReaderModule.cpp

```diff
@@ -62,5 +62,87 @@ void MagneticFieldReaderModule::initialize() {
                        << Units::display(detector->getMagneticField(detector->getLocalPosition(position)), {"T", "mT"});
         }
         LOG(INFO) << "Set constant magnetic field: " << Units::display(b_field, {"T", "mT"});
+    } else if(field_model == MagneticField::MESH) {
+        type = MagneticFieldType::CUSTOM;
+
+        auto fallback_field = config_.get<ROOT::Math::XYZVector>("magnetic_field_fallback", ROOT::Math::XYZVector());
+        auto field_data = read_field();
+
+        auto dims = field_data.getDimensionality(); // Vector dimensionality of the field (should be 3 for B-field)
+        if(dims != 3) {
+            throw std::invalid_argument("B-field vectors were not three dimensional.");
+        }
+
+        auto cellsize = field_data.getSize();     // Physical dimensions for each cell in the field mesh
+        auto ncells = field_data.getDimensions(); // Number of cells in each direction of the field mesh
+        auto field_mesh = field_data.getData();   // The field mesh B-field data as a flattened array
+
+        if(ncells[0] * ncells[1] * ncells[2] * dims != field_mesh->size()) {
+            throw std::invalid_argument("B-field does not span the given dimensions");
+        }
+
+        MagneticFieldFunction function =
+            [fallback_field, field_mesh, cellsize, ncells, dims](const ROOT::Math::XYZPoint& coord) {
+                // Find the nearest field mesh cell to the given input coordinate (assuming the mesh is centred about the
+                // origin)
+                auto xind = std::round((coord.X() / cellsize[0]) + (static_cast<double>(ncells[0]) / 2.));
+                auto yind = std::round((coord.Y() / cellsize[1]) + (static_cast<double>(ncells[1]) / 2.));
+                auto zind = std::round((coord.Z() / cellsize[2]) + (static_cast<double>(ncells[2]) / 2.));
+
+                if(xind < 0 || xind >= static_cast<double>(ncells[0]) || yind < 0 ||
+                   yind >= static_cast<double>(ncells[1]) || zind < 0 || zind >= static_cast<double>(ncells[2])) {
+                    return fallback_field;
```

Comment:
Instead of using a fallback field value, I probably would simply `std::clamp` the indices to the valid range and thereby using the last valid point at the field edge of the defined field.

This makes for a simpler user interface and removes potential and unexpected steps at the boundary.

Reply:
A case where the fallback value could be useful is for a setup placed in, say, a region of constant magnetic field, but that contains some geometry with shielding, that therefore has a different local magnetic field. Here, the meshed magnetic field could be used to specify the shielded area's field, with the fallback setting a (potentially quite different) field value for the rest of the world volume.

Perhaps clamping the indices to the valid range should be used if no fallback field is set?

Reply:
I see, but wouldn't this shielded field still have to be somehow continuous and have a transition tot he "world" field around it? Probably I'm overthinking it. :)